### ISSUE TITLE
`dfs_reach` does not visit nodes in `successors` order

### ISSUE THREAD

**Issue by stepancheg** [evenfurther/pathfinding#511]
Test

```rust
#[test]
fn dfs_reach_bug_q() {
    let mut it = dfs_reach(0, |&n| [n + 1, n + 5].into_iter().filter(|&x| x <= 10));
    assert_eq!(vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 10], it.collect::<Vec<u32>>());
}
```

error:

```
Left:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Right: [0, 1, 2, 3, 4, 9, 10, 8, 7, 6, 5]
```

I think after visiting node `4` we should go deep to node `5`.

That would be consistent with this [example from Wikipedia](https://en.wikipedia.org/wiki/Depth-first_search#Example).

**Comment by samueltardieu** [comment]
On the one hand, this is DFS because it does a depth first search. On the other hand, we do indeed bring the guarantee that nodes are visited in the order returned by the `successors` function in the [documentation](https://docs.rs/pathfinding/4.8.1/pathfinding/directed/dfs/fn.dfs_reach.html).

So this is indeed a bug. I'll rename the issue, thanks for reporting!

**Comment by samueltardieu** [comment]
This is fixed in pathfinding 4.8.2 which has just been released.
